wallets/metamask/src/utils/waitFor.ts

```diff
@@ -21,6 +22,60 @@ export const waitForSelector = async (selector: string, page: Page, timeout: num
   }
 }
 
+export const waitForMetaMaskLoad = async (page: Page) => {
+  await Promise.all(
+    LoadingSelectors.loadingIndicators.map(async (selector) => {
+      await waitForSelector(selector, page, 5000)
+    })
+  )
+    .then(() => {
+      console.log('All loading indicators are hidden')
+    })
+    .catch((error) => {
+      console.error('Error: ', error)
+    })
+
+  return page
+}
+
+export const waitForMetaMaskWindowToBeStable = async (page: Page) => {
+  await waitForMetaMaskLoad(page)
+  if ((await page.locator(ErrorSelectors.loadingOverlayErrorButtons).count()) > 0) {
+    const retryButton = await page.locator(ErrorSelectors.loadingOverlayErrorButtonsRetryButton)
+    await retryButton.click()
+    await waitForSelector(LoadingSelectors.loadingOverlay, page, 300)
+  }
+  await fixCriticalError(page)
+}
+
+export const fixCriticalError = async (page: Page) => {
+  for (let times = 0; times < 5; times++) {
+    if ((await page.locator(ErrorSelectors.criticalError).count()) > 0) {
+      console.log('[fixCriticalError] Metamask crashed with critical error, refreshing..')
+      if (times <= 3) {
+        await page.reload()
+        await waitForMetaMaskWindowToBeStable(page)
+      } else if (times === 4) {
+        const restartButton = await page.locator(ErrorSelectors.criticalErrorRestartButton)
+        await restartButton.click()
+        await waitForMetaMaskWindowToBeStable(page)
+      } else {
+        throw new Error('[fixCriticalError] Max amount of retries to fix critical metamask error has been reached.')
```

Comment:
how we define 5 as max?

Reply:
It's an arbitrary number we defined in V3 it could be anything but 5 seems to be enough.